autogpt_platform/backend/backend/blocks/human_in_the_loop.py

```diff
@@ -0,0 +1,147 @@
+                ("review_message", ""),
+            ],
+            test_mock={
+                "handle_review_workflow": lambda *args, **kwargs: HumanInTheLoopBlock._create_test_result(),
+                "update_graph_execution_stats": lambda *args, **kwargs: None,
+            },
+        )
+
+    @staticmethod
+    def _create_test_result():
+        """Create test result for mocking"""
+        from backend.blocks.human_in_the_loop_service import ReviewResult
+
+        return ReviewResult(
+            data={"name": "John Doe", "age": 30}, status="approved", message=""
+        )
+
+    async def handle_review_workflow(self, *args, **kwargs):
+        """Wrapper method for HumanInTheLoopService.handle_review_workflow that can be mocked"""
+        return await HumanInTheLoopService.handle_review_workflow(*args, **kwargs)
+
+    async def update_graph_execution_stats(self, *args, **kwargs):
+        """Wrapper method for update_graph_execution_stats that can be mocked"""
+        return await update_graph_execution_stats(*args, **kwargs)
+
+    async def run(
+        self,
+        input_data: Input,
+        *,
+        user_id: str,
+        node_exec_id: str,
+        graph_exec_id: str,
+        graph_id: str,
+        graph_version: int,
+        **kwargs,
+    ) -> BlockOutput:
+        """
+        Execute the Human In The Loop block.
+
+        This method uses the HumanInTheLoopService to handle all business logic,
+        keeping the block implementation clean and focused on data flow.
+        """
+        # Use the service to handle the complete workflow
+        result = await self.handle_review_workflow(
+            user_id=user_id,
+            node_exec_id=node_exec_id,
+            graph_exec_id=graph_exec_id,
+            graph_id=graph_id,
+            graph_version=graph_version,
+            input_data=input_data.data,
+            message=input_data.message,
+            editable=input_data.editable,
+            expected_data_type=type(input_data.data),
+        )
+
+        if result is not None:
+            # Review is complete (approved or rejected)
+            if result.status == "approved":
+                yield "reviewed_data", result.data
+                yield "status", "approved"
+                yield "review_message", result.message
+            elif result.status == "rejected":
+                yield "status", "rejected"
+                yield "review_message", result.message
+                # Raise an exception for rejected reviews to stop execution
+                raise HITLValidationError(
+                    f"Human review rejected: {result.message}", result.message
+                )
+            return
+
+        # No result means we're waiting for human input
+        # Update the graph execution status to indicate waiting state
+        await self.update_graph_execution_stats(
+            graph_exec_id=graph_exec_id, status=ExecutionStatus.WAITING_FOR_REVIEW
+        )
```

Comment:
if we update the status here, then when the node execution complete we will still marking it as COMPELTED and this will be futile

Reply:
the block won't wait, it's just completed with the WAITING_FOR_REVIEW state, then there is another api that will retrigger the graph execution after approving the data which then we should also requeue the graph execution with  inWAITING_FOR_REVIEW added into the node queue, and when this is executed atain the block will check whether we already have data for this node_Exec if we do we set the status to complete and we are good.